src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,303 @@
+    ///         if (n &lt;= 2)
+    ///             return 1
+    ///             
+    ///         return fib(n - 1) + fib(n - 2);
+    ///     }
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpUseLocalFunctionDiagnosticAnalyzer : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public CSharpUseLocalFunctionDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseLocalFunctionDiagnosticId,
+                   new LocalizableResourceString(
+                       nameof(FeaturesResources.Use_local_function), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(SyntaxNodeAction, SyntaxKind.SimpleLambdaExpression, SyntaxKind.ParenthesizedLambdaExpression);
+
+        private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
+        {
+            var options = syntaxContext.Options;
+            var syntaxTree = syntaxContext.Node.SyntaxTree;
+            var cancellationToken = syntaxContext.CancellationToken;
+            var optionSet = options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var styleOption = optionSet.GetOption(CSharpCodeStyleOptions.PreferLocalOverAnonymousFunction);
+            if (!styleOption.Value)
+            {
+                // Bail immediately if the user has disabled this feature.
+                return;
+            }
+
+            // Local functions are only available in C# 7.0 and above.  Don't offer this refactoring
+            // in projects targetting a lesser version.
+            if (((CSharpParseOptions)syntaxTree.Options).LanguageVersion < LanguageVersion.CSharp7)
+            {
+                return;
+            }
+
+            var severity = styleOption.Notification.Value;
+            var anonymousFunction = (AnonymousFunctionExpressionSyntax)syntaxContext.Node;
+
+            var semanticModel = syntaxContext.SemanticModel;
+            if (!CheckForPattern(semanticModel, anonymousFunction, cancellationToken,
+                    out var localDeclaration))
+            {
+                return;
+            }
+
+            if (localDeclaration.Declaration.Variables.Count != 1)
+            {
+                return;
+            }
+
+            if (!(localDeclaration.Parent is BlockSyntax block))
+            {
+                return;
+            }
+
+            var local = semanticModel.GetDeclaredSymbol(localDeclaration.Declaration.Variables[0], cancellationToken);
+            if (local == null)
+            {
+                return;
+            }
+
+            if (IsWrittenAfter(semanticModel, local, block, anonymousFunction, cancellationToken))
```

Comment:
What if the local is *passed* to something that takes a lambda?  I guess the language will automatically create one at this point?  Is there a problem with delegate co/contra-variance in the casted case?